panel/widgets/chat.py

```diff
@@ -0,0 +1,1328 @@
+        """
+        Toggle the button's icon and name to indicate
+        whether the action can be reverted.
+        """
+        for button in button_data.buttons:
+            if active and button_data.objects:
+                button.button_type = "warning"
+                button.name = "Revert"
+                button.max_width = 75
+            else:
+                button.button_type = "default"
+                if self.show_button_name:
+                    button.name = button_data.name.title()
+                else:
+                    button.name = ""
+                button.max_width = 45
+
+    def _reset_button_data(self):
+        """
+        Clears all the objects in the button data
+        to prevent reverting.
+        """
+        for button_data in self._button_data.values():
+            button_data.objects.clear()
+            self._toggle_revert(button_data, False)
+
+    def _click_rerun(self, _):
+        """
+        Upon clicking the rerun button, rerun the last user entry,
+        which can trigger the callback again.
+        """
+        count = self._get_last_user_entry_index()
+        entries = self.undo(count)
+        if not entries:
+            return
+        self.send(value=entries[0], respond=True)
+
+    def _click_undo(self, _):
+        """
+        Upon clicking the undo button, undo (remove) entries
+        up to the last user entry. If the button is clicked
+        again without performing any other actions, revert the undo.
+        """
+        undo_data = self._button_data["undo"]
+        undo_objects = undo_data.objects
+        if not undo_objects:
+            self._reset_button_data()
+            count = self._get_last_user_entry_index()
+            undo_data.objects = self.undo(count)
+            self._toggle_revert(undo_data, True)
+        else:
+            self.value = [*self.value, *undo_objects.copy()]
+            self._reset_button_data()
+
+    def _click_clear(self, _):
+        """
+        Upon clicking the clear button, clear the chat log.
+        If the button is clicked again without performing any
+        other actions, revert the clear.
+        """
+        clear_data = self._button_data["clear"]
+        clear_objects = clear_data.objects
+        if not clear_objects:
+            self._reset_button_data()
+            clear_data.objects = self.clear()
+            self._toggle_revert(clear_data, True)
+        else:
+            self.value = clear_objects.copy()
+            self._reset_button_data()
+
+    @property
+    def active_widget(self) -> Widget:
+        """
+        The currently active widget.
+
+        Returns
+        -------
+        The active widget.
+        """
+        if isinstance(self._input_layout, Tabs):
+            active = self._input_layout.active
+            return self._input_layout[active].objects[0]
+        else:
+            return self._input_layout.objects[0]
+
+    @property
+    def active_tab(self) -> int:
```

Comment:
I believe we can make this a bit more general and future proof if we rename this to something like `active`, `active_index` or `widget_index` or `active_widget_index`.

`active` would be in line with Tabs.

This will enable us to change the layout from `Tab` to something else in the future. The user cares about which `widget` or input is active. Not which `tab`.

Please consider whether this makes sense? Thanks.